File: backend/app/models/scan.py
import json
from datetime import datetime
from enum import IntEnum
from typing import TYPE_CHECKING, List, Optional

from pydantic import computed_field
from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from .result import Result
    from .user import User


class UserScanLink(SQLModel, table=True):
    user_id: Optional[int] = Field(
        default=None, foreign_key="user.id", primary_key=True
    )
    scan_id: Optional[int] = Field(
        default=None, foreign_key="scan.id", primary_key=True
    )


class ScanStatus(IntEnum):
    pending = 0
    running = 1
    completed = 2
    failed = 3


class ScanType(IntEnum):
    simple = 0
    complex = 1


class Scan(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    url: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    status: ScanStatus = ScanStatus.pending
    progress: int = 0
    type: ScanType
    current_step: str = ""
    data: str = "{}"

    # Creator relationship (one-to-many)
    creator_id: Optional[int] = Field(default=None, foreign_key="user.id")
    creator: Optional["User"] = Relationship(
        back_populates="created_scans",
        sa_relationship_kwargs={"foreign_keys": "[Scan.creator_id]"},
    )

    # Many-to-many relationship with users
    users: List["User"] = Relationship(back_populates="scans", link_model=UserScanLink)

    # One to many relationship with results
    results: List["Result"] = Relationship(back_populates="scan")

    @property
    def data_dict(self):
        return json.loads(self.data)

    @data_dict.setter
    def data_dict(self, value):
        self.data = json.dumps(value)

    @computed_field
    @property
    def score(self) -> float:
        """Returns the mean of all result scores ignoring those set to -1."""
        scores = [result.score for result in self.results if result.score != -1]
        return sum(scores) / len(scores) if scores else 0

File: frontend/src/types/scan.ts
export enum ScanStatus {
  pending = 0,
  running = 1,
  completed = 2,
  failed = 3,
}

export enum ScanType {
  simple = 0,
  complex = 1,
}

export interface Scan {
  id: number;
  type: ScanType;
  url: string;
  status: ScanStatus;
  created_at: string;
  progress: number;
  current_step: string;
  data: string;
  creator_id: number;
  score: number | null;
}

export const getScanTypeText = (type: ScanType): string => {
  switch (type) {
    case ScanType.simple:
      return "Simple";
    case ScanType.complex:
      return "Complex";
    default:
      return "Unknown";
  }
};

export const getScanStatusText = (status: ScanStatus): string => {
  switch (status) {
    case ScanStatus.pending:
      return "Pending";
    case ScanStatus.running:
      return "Running";
    case ScanStatus.completed:
      return "Completed";
    case ScanStatus.failed:
      return "Failed";
    default:
      return "Unknown";
  }
};

export const getScanStatusClass = (status: ScanStatus): string => {
  switch (status) {
    case ScanStatus.pending:
      return "badge-warning";
    case ScanStatus.running:
      return "badge-info";
    case ScanStatus.completed:
      return "badge-success";
    case ScanStatus.failed:
      return "badge-error";
    default:
      return "badge-ghost";
  }
};

export interface Result {
  id: number;
  title: string;
  severity: number;
  category: number;
  description: string;
  short_description: string;
  recommendation: string;
  score: number;
  scan_id: number;
}
